### PR TITLE
Require login to view job and project data

### LINKED ISSUE #122: Add token or password
### Feature description

Require the use of token or password security **by default**.

### Proposed solution

For example, (by default) `jupyter lab` creates a random token and displays it on the console.

### Additional context

Without authentication, `signac-dashboard` instances are open to all users on a multi-user system (such as a cluster login node). Malicious users could use this connection to read your signac dashboard, write if any modules enable writing, and possibly take control of your account through vulnerabilities.

### PR DESCRIPTION
**Pull request by joaander**
## Description

Following the model of Jupyter, generate a token when launched. Users must login with that token before they can view sensitive data. All routes are protected, except for static content.

This feature requires an additional dependency: `flask-login`: https://flask-login.readthedocs.io/ which appears to be the best maintained package for the job.

## Motivation and Context

On multi-user systems, anyone with access to the system can connect to servers running on open ports. This opens Dashboard user's projects to be read (and written to in the case of some modules) by anyone with access to the host system. 

This pull requests prevents that access. Like Jupyter, it generates a login token when invoked on the command line. Following the security guidelines in the documentation, users will receive this token over a secure SSH channel.
```
$ python3 dashboard.py run
/home/joaander/venv/lib/python3.10/site-packages/signac/contrib/project.py:95: FutureWarning: This method will soon become a property and should be used without the call operator (parentheses).
  warnings.warn(
To access this server, connect to: http://localhost:8888/login?token=<hidden>
 * Serving Flask app 'signac-dashboard'
 * Debug mode: off
WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://localhost:8888
Press CTRL+C to quit
```

Other users that attempt to access the dashboard are greeted with:
image

In future work, we could add username/password authentication, where the user provides a hashed password in a config file.

Fixes #122 

## Checklist:

- [x] I am familiar with the [Contributing Guidelines](https://github.com/glotzerlab/signac-dashboard/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [Contributor Agreement](https://github.com/glotzerlab/signac-dashboard/blob/master/ContributorAgreement.md).
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.
- [ ] The [package documentation](https://github.com/glotzerlab/signac-dashboard/tree/master/doc) and [framework documentation](https://docs.signac.io/) in [signac-docs](https://github.com/glotzerlab/signac-docs) are up to date with these changes.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac-dashboard/blob/master/changelog.txt) and added any related issue and pull request numbers for future reference.

## Documentation:

I will submit a request to update the framework documentation after this request is approved.

### CODE REVIEW

File: changelog.txt [hunk 1]
```diff
@@ -7,6 +7,15 @@ The **signac-dashboard** package follows `semantic versioning <https://semver.or
...
+[next release] -- YYYY-MM-DD
+----------------------------
+
+Changed
++++++++
+
+- Improve security on multi-user systems. Dashboard now generates a login token when started. Users
+  must login with the token to view project and job data in the dashboard.
+
 [0.3.1] -- 2022-10-17
```
Comment (bdice) [inline changelog.txt:19]: We missed this in the last release.
```suggestion
[0.3.1] -- 2022-10-17
---------------------
```

File: changelog.txt [hunk 2]
```diff
@@ -7,6 +7,15 @@ The **signac-dashboard** package follows `semantic versioning <https://semver.or
 Version 0.3
 ===========
 
+[next release] -- YYYY-MM-DD
```
Comment (bdice) [inline changelog.txt:10]: I'd put this in a `Version 0.4` section. This feels worthy of a minor version bump.

File: changelog.txt [hunk 3]
```diff
@@ -7,6 +7,15 @@ The **signac-dashboard** package follows `semantic versioning <https://semver.or
...
+----------------------------
+
+Changed
++++++++
+
+- Improve security on multi-user systems. Dashboard now generates a login token when started. Users
```
Comment (bdice) [inline changelog.txt:16]: This feature is probably "Added" more than "Changed".

File: signac_dashboard/dashboard.py [hunk 4]
```diff
@@ -105,9 +121,24 @@ def __init__(self, config={}, project=None, modules=[]):
                 f"{self.config['CARDS_PER_ROW']}."
             )
 
+        if "ACCESS_TOKEN" not in config:
+            access_token = secrets.token_urlsafe()
+            self.config["ACCESS_TOKEN"] = access_token
```
Comment (bdice) [inline signac_dashboard/dashboard.py:126]: I think this is equivalent.
```suggestion
        self.config.setdefault("ACCESS_TOKEN", secrets.token_urlsafe())
```

File: signac_dashboard/dashboard.py [hunk 5]
```diff
@@ -548,7 +579,22 @@ def modify_query(**new_values):
         def page_not_found(error):
             return self._render_error(str(error))
 
+        @dashboard.login_manager.unauthorized_handler
+        def unauthorized_handler():
+            return self._render_error("Unauthorized")
```
Comment (bdice) [inline signac_dashboard/dashboard.py:584]: Perhaps something more verbose? Maybe "Access token is required."

**Review by bdice — COMMENTED** [review]
The code in this PR looks broadly correct, from my (limited) experience with web login systems. Thank you @joaander!

File: changelog.txt (hunk 2 above)
Comment (joaander) [reply changelog.txt:10]: Will do.

**Comment by joaander** [comment]
> Should it be possible to disable the token-based login? e.g. if `ACCESS_TOKEN is None`.

This would make it more convenient for users on systems such as OLCF Marble, which already adds a layer of security. We could make this possible and warn users about it in the security page.

File: signac_dashboard/dashboard.py [hunk 6]
```diff
@@ -548,7 +589,22 @@ def modify_query(**new_values):
...
+        def unauthorized_handler():
+            return self._render_error("Access token is required.")
+
+        @dashboard.app.route("/login")
+        def login():
+            provided_token = request.args.get("token")
+            if provided_token == self.config["ACCESS_TOKEN"]:
+                user = User(provided_token)
+                flask_login.login_user(user)
+                return redirect("/")
```
Comment (cbkerr) [inline signac_dashboard/dashboard.py:602]: Should we allow the user to provide the token when visiting any route?
(Is that possible?)

File: signac_dashboard/util.py [hunk 7]
```diff
@@ -56,5 +57,8 @@ def view(self):
         return import_string(self.import_name)
 
     def __call__(self, *args, **kwargs):
+        if not flask_login.current_user.is_authenticated:
+            return self.dashboard.login_manager.unauthorized(), 401
```
Comment (cbkerr) [inline signac_dashboard/util.py:61]:
```suggestion
        # Protect routes added in the format self.add_url("views.home", ["/"])
        if not flask_login.current_user.is_authenticated:
            return self.dashboard.login_manager.unauthorized(), 401
```

File: tests/test_dashboard.py [hunk 8]
```diff
@@ -150,6 +163,11 @@ def setUp(self):
         self.test_client = self.dashboard.app.test_client()
         self.addCleanup(shutil.rmtree, self._tmp_dir)
 
+    def test_login(self):
```
Comment (cbkerr) [inline tests/test_dashboard.py:166]:
```suggestion
    def test_login_with_None_token(self):
```

File: signac_dashboard/dashboard.py [hunk 9]
```diff
@@ -548,10 +589,26 @@ def modify_query(**new_values):
...
+
+            return self._render_error("Invalid token")
+
         @dashboard.app.route("/favicon.ico")
+        @flask_login.login_required
         def favicon():
             return url_for("static", filename="favicon.ico")
 
+        # These routes are protected within the LazyView utility class
```
Comment (cbkerr) [inline signac_dashboard/dashboard.py:611]: I wanted to point people to what you added in the `__call__` function of `LazyView`

File: .github/workflows/ci-oldest-reqs.txt [hunk 10]
```diff
@@ -12,3 +12,4 @@ signac==1.0.0
 watchdog
 webassets==2.0.0
 werkzeug==2.1.0
+flask-login==0.6.0
```
Comment (bdice) [inline .github/workflows/ci-oldest-reqs.txt:15]: Alphabetize this list.

File: requirements.txt [hunk 11]
```diff
@@ -10,3 +10,4 @@ signac>=1.0.0
 watchdog
 webassets>=2.0.0
 werkzeug>=2.1.0
+flask-login>=0.6.0
```
Comment (bdice) [inline requirements.txt:13]: Alphabetize.

File: setup.py [hunk 12]
```diff
@@ -18,6 +18,7 @@
     "watchdog",
     "webassets>=2.0.0",
     "werkzeug>=2.1.0",
+    "flask-login>=0.6.0",
```
Comment (bdice) [inline setup.py:21]: Alphabetize.

File: signac_dashboard/dashboard.py [hunk 13]
```diff
@@ -63,6 +76,9 @@ class such as :code:`MyDashboard` which can be customized and launched via
       (default: 25).
     - **CARDS_PER_ROW**: Cards to show per row in the desktop view. Must be a
       factor of 12 (default: 3).
+    - **ACCESS_TOKEN**: The access token required to login to the dashboard.
+      Set to `None` to disable authentication (not recommended on multi-user
```
Comment (bdice) [inline signac_dashboard/dashboard.py:80]:
```suggestion
      Set to :code:`None` to disable authentication (not recommended on multi-user
```

File: signac_dashboard/dashboard.py [hunk 14]
```diff
@@ -105,9 +123,32 @@ def __init__(self, config={}, project=None, modules=[]):
...
         self.app = self._create_app(self.config)
 
+        # Initialize the login manager
+        self.login_manager = flask_login.LoginManager()
+        self.login_manager.init_app(self.app)
+
+        @self.login_manager.user_loader
+        def user_loader(identifier):
+            if self.config["ACCESS_TOKEN"] is None:
+                return User("None")
```
Comment (bdice) [inline signac_dashboard/dashboard.py:138]: Must the `id` be a string? If not, I'd prefer `User(None)`.

**Review by bdice — APPROVED** [review]
This looks good at a high level. I haven't tested it locally yet but I am fine with this PR being merged if others have tested it.

File: signac_dashboard/dashboard.py [hunk 15]
```diff
@@ -548,7 +589,22 @@ def modify_query(**new_values):
         def page_not_found(error):
             return self._render_error(str(error))
 
+        @dashboard.login_manager.unauthorized_handler
+        def unauthorized_handler():
+            return self._render_error("Access token is required.")
```
[the remark this answers is not in the document]
Comment (joaander) [reply signac_dashboard/dashboard.py:594]: Agreed.

File: signac_dashboard/dashboard.py (hunk 14 above)
Comment (joaander) [reply signac_dashboard/dashboard.py:138]: Yes, the user id must be a `str`: https://flask-login.readthedocs.io/en/latest/#your-user-class
> Note that this must be a [str](https://docs.python.org/3/library/stdtypes.html#str) - if the ID is natively an [int](https://docs.python.org/3/library/functions.html#int) or some other type, you will need to convert it to [str](https://docs.python.org/3/library/stdtypes.html#str).

File: signac_dashboard/dashboard.py (hunk 6 above)
Comment (joaander) [reply signac_dashboard/dashboard.py:602]: Yes, it is possible. No, I don't think we should allow it.

**Comment by joaander** [comment]
> This looks good at a high level. I haven't tested it locally yet but I am fine with this PR being merged if others have tested it.

I have tested it. @cbkerr, have you?

File: signac_dashboard/dashboard.py (hunk 14 above)
Comment (bdice) [reply signac_dashboard/dashboard.py:138]: Thanks! 👍 Sorry, I hadn’t looked at the flask-login docs at all.

**Comment by bdice** [comment]
@joaander @cbkerr I think this can merge when you are ready.

**Comment by cbkerr** [comment]
I have run it locally with my test dashboard and a test signac project with the Statepoint, ImageViewer, JobDocument, ProjectSchema, Notes and Document Editor modules, but not all modules.

**Pull request merged**